evm/src/cpu/kernel/asm/curve/bn254/field_arithmetic/util.asm

```diff
@@ -145,253 +164,259 @@
-    // stack:                   ind2, ptr
-    %mload_bn254_pairing
-    // stack:                     x2, ptr
-    DUP2  
+    // stack:                  addr2, addr
+    MLOAD_GENERAL
+    // stack:                     x2, addr
+    DUP2
     %add_const(1)
-    // stack:               ind1, x2, ptr
-    %mload_bn254_pairing
-    // stack:                 x1, x2, ptr
-    DUP3  
+    // stack:              addr1, x2, addr
+    MLOAD_GENERAL
+    // stack:                 x1, x2, addr
+    DUP3
     %add_const(3)
-    // stack:           ind3, x1, x2, ptr
-    %mload_bn254_pairing
-    // stack:             x3, x1, x2, ptr
+    // stack:          addr3, x1, x2, addr
+    MLOAD_GENERAL
+    // stack:             x3, x1, x2, addr
     SWAP3
-    // stack:            ind0, x1, x2, x3
-    %mload_bn254_pairing
+    // stack:           addr0, x1, x2, x3
+    MLOAD_GENERAL
     // stack:              x0, x1, x2, x3
 %endmacro
 
 // fp254_6 macros
 
 %macro load_fp254_6
     // stack:                         ptr
-    DUP1  
+    %create_bn254_pairing_address
+    DUP1
     %add_const(4)
-    // stack:                   ind4, ptr
-    %mload_bn254_pairing
-    // stack:                     x4, ptr
-    DUP2  
+    // stack:                   addr4, addr
+    MLOAD_GENERAL
+    // stack:                     x4, addr
+    DUP2
     %add_const(3)
-    // stack:               ind3, x4, ptr
-    %mload_bn254_pairing
-    // stack:                 x3, x4, ptr
-    DUP3  
+    // stack:               addr3, x4, addr
+    MLOAD_GENERAL
+    // stack:                 x3, x4, addr
+    DUP3
     %add_const(2)
-    // stack:           ind2, x3, x4, ptr
-    %mload_bn254_pairing
-    // stack:             x2, x3, x4, ptr
-    DUP4  
+    // stack:           addr2, x3, x4, addr
+    MLOAD_GENERAL
+    // stack:             x2, x3, x4, addr
+    DUP4
     %add_const(1)
-    // stack:       ind1, x2, x3, x4, ptr
-    %mload_bn254_pairing
-    // stack:         x1, x2, x3, x4, ptr
-    DUP5  
+    // stack:       addr1, x2, x3, x4, addr
+    MLOAD_GENERAL
+    // stack:         x1, x2, x3, x4, addr
+    DUP5
     %add_const(5)
-    // stack:   ind5, x1, x2, x3, x4, ptr
-    %mload_bn254_pairing
-    // stack:     x5, x1, x2, x3, x4, ptr
+    // stack:   addr5, x1, x2, x3, x4, addr
+    MLOAD_GENERAL
+    // stack:     x5, x1, x2, x3, x4, addr
     SWAP5
-    // stack:   ind0, x1, x2, x3, x4, x5
-    %mload_bn254_pairing
+    // stack:   addr0, x1, x2, x3, x4, x5
+    MLOAD_GENERAL
     // stack:     x0, x1, x2, x3, x4, x5
 %endmacro
 
 // cost: 6 loads + 6 pushes + 5 adds = 6*4 + 6*1 + 5*2 = 40
 %macro load_fp254_6(ptr)
     // stack:
-    PUSH $ptr  
+    PUSH $ptr
+    %create_bn254_pairing_address
+    DUP1
     %add_const(5)
-    // stack:                     ind5
-    %mload_bn254_pairing
-    // stack:                       x5
-    PUSH $ptr  
+    // stack:                 addr5, addr
+    MLOAD_GENERAL
+    // stack:                    x5, addr
```

Comment:
Why not follow the same approach as for the macro above (`load_fp254_6`): load x1 to x4 using only DUPs (leaving addr at the back of the stack), then load x5, then SWAP and load x0?

Reply:
The other macro being cheaper, I just replaced it with 
```asm
%macro load_fp254_6(ptr)
    // stack:
    PUSH $ptr
    %load_fp254_6
    // stack: x0, x1, x2, x3, x4, x5
%endmacro
```
instead

Reply:
The starting ptr isn't at the back of the stack but at the start, hence the difference with `load_fp254_6`

Reply:
I'm sorry, I'm not sure I understand: I meant that at this point, the address is at the back so for the rest of the macro, instead of having a SWAP and DUP every time, why not use the trick above?

Reply:
Hmm I'm confused, the address is on top of the stack here

Reply:
The stack is: `x5, addr`, right?